### PR TITLE
fix: abort ShipIt installation attempt at the final mile if the app is running

### PR DESCRIPTION
**Pull request by MarshallOfSound**
As in title, basically there is a race condition in Squirrel.Mac.  The current flow looks like this

```
1. Download update
2. Quit app and launch ShipIt
3. ShipIt waits for app to terminate
4. ShipIt prepares update, unzips, validates, etc.
5. ShipIt installs update
6. ShipIt launches new app
```

If _somehow_ the app manages to launch (the user clicks it, programatic relaunch, etc.) **during** step 4 then step 5 succeeds, step 6 bails as the app is running and everything _appears_ to be fine.  **But it is not....** The running app no longer exists on disk and will therefore fail all uncached TCC / Security.h checks including microphone/camera permission checks and keychain access checks.  (Amongst other awful things).

This fixes this race condition by adding a step 4.5 that checks if any app is running from the bundle path we're about to install an update to.  We don't do step (3) again because waiting for the app to quit is unrealistic if the user manually launched it.  The best thing to do is to get back into a known good state and bail the update flow with an error but without triggering an automatic launchd retry (hence the "Happy error" exit logic).

Notes: Fixed race condition during update on macOS that could result in TCC and Keychain errors

### CODE REVIEW

File: spec/yarn.lock [hunk 1]
```diff
@@ -303,7 +302,7 @@ dashdash@^1.12.0:
     safe-buffer "^5.1.1"
     xml2js "^0.4.17"
   optionalDependencies:
-    abstract-socket "^2.0.0"
+    abstract-socket "github:saghul/node-abstractsocket#35b1b1491fabc04899bde5be3428abf5cf9cd528"
```
Comment (nornagon) [inline spec/yarn.lock:305]: is this related?

File: patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ NSString * const SQRLInstallerOwnedBundleKey = @"SQRLInstallerOwnedBundle";
+@@ -286,6 +287,19 @@ - (RACSignal *)installRequest:(SQRLShipItRequest *)request {
+ 			return [[[[self
+ 				renameIfNeeded:request updateBundleURL:updateBundleURL]
+ 				flattenMap:^(SQRLShipItRequest *request) {
++					// Final validation that the application is not running again;
++					NSArray *apps = [[NSRunningApplication runningApplicationsWithBundleIdentifier:request.bundleIdentifier] filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(NSRunningApplication *app, NSDictionary *bindings) {
++						return [[[app bundleURL] URLByStandardizingPath] isEqual:request.targetBundleURL];
++					}]];
++					if ([apps count] != 0) {
```
Comment (nornagon) [inline patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch:43]: a) how come the bundleIdentifier _and_ targetBundleURL must match? (what's the difference between those two?)
b) `filteredArrayUsingPredicate:[NSPredicate predicateWithBlock` is a really verbose way to write `filter`, eh? 😅 maybe [indexOfObjectPassingTest:](https://developer.apple.com/documentation/foundation/nsarray/1408043-indexofobjectpassingtest?language=objc)?

File: patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+--- a/Squirrel/ShipIt-main.m
++++ b/Squirrel/ShipIt-main.m
+@@ -201,8 +201,14 @@ static void installRequest(RACSignal *readRequestSignal, NSString *applicationId
+ 			return action;
+ 		}]
+ 		subscribeError:^(NSError *error) {
+-			NSLog(@"Installation error: %@", error);
+-			exit(EXIT_FAILURE);
++			if ([[error domain] isEqual:SQRLInstallerErrorDomain] && [error code] == SQRLInstallerErrorAppStillRunning) {
++				NSLog(@"Installation error but happy: %@", error);
```
Comment (nornagon) [inline patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch:66]: let's make this a bit more descriptive?
```suggestion
+				NSLog(@"Installation cancelled: %@", error);
```

**Review by nornagon — APPROVED** [review]
:shipit:

File: spec/yarn.lock (hunk 1 above)
Comment (MarshallOfSound) [reply spec/yarn.lock:305]: Uh, no.  I just ran `yarn`.  Not sure why this wasn't accurate before. I'll double check nothing nefarious is afoot

File: patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch (hunk 2 above)
Comment (MarshallOfSound) [reply patches/squirrel.mac/fix_abort_installation_attempt_at_the_final_mile_if_the_app_is.patch:43]: `bundleIdentifier` === `com.github.Electron`
`bundleUrl` === `file:///Applications/Electron.app`

We care if both match because that's the only case that would cause issues

File: spec/yarn.lock (hunk 1 above)
Comment (MarshallOfSound) [reply spec/yarn.lock:305]: Yeah this checks out, not sure why this wasn't updated before

**Review by VerteDinde — APPROVED** [review]

**Review by zcbenz — APPROVED** [review]

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Newly added test fails
